cmd/gb-vendor/vendor/repo.go

```diff
@@ -137,13 +134,22 @@ func (g *gitrepo) URL() string {
 	return g.url
 }
 
-func (g *gitrepo) Clone() (WorkingCopy, error) {
+func (g *gitrepo) Checkout(branch, revision string) (WorkingCopy, error) {
 	dir, err := mktmp()
 	if err != nil {
 		return nil, err
 	}
 
-	if err := runOut(os.Stderr, "git", "clone", g.url, dir); err != nil {
+	args := []string{
+		"clone",
+		g.url,
+		dir,
+	}
+	if branch != "" {
+		args = append(args, "--branch", branch)
```

Comment:
Well honestly I would remove the support for revisions completely. Anyone who checks out revisions in projects, either does not know how to use vcs or wants to do something completely silly which can be done better (of course there could be some cases where this could be needed, but I have never seen such).

That would also remove the problem with `Detached HEAD`. 

Nevertheless, you can checkout either branch or commit. You can't do both at once. User can't specify branch and revision at once, there must be a check if user didn't do that. 

If you want to checkout commit (revision) you can do something like this:
```bash
git clone $repo
git reset --hard $revision
```